linkerd/tls/src/client.rs

```diff
@@ -85,26 +109,45 @@ where
     }
 
     fn call(&mut self, target: T) -> Self::Future {
-        let tls = match self.local.as_ref() {
-            Some(l) => tokio_rustls::TlsConnector::from(l.into()),
-            None => {
-                trace!("Local identity disabled");
+        let ClientTls { server_id, alpn } = match (&target).into() {
+            Conditional::Some(tls) => tls,
+            Conditional::None(reason) => {
+                debug!(%reason, "Peer does not support TLS");
                 return Either::Left(self.inner.call(target).map_ok(io::EitherIo::Left));
             }
         };
-        let server_id = match (&target).into() {
-            Conditional::Some(ServerId(id)) => id,
-            Conditional::None(reason) => {
-                debug!(%reason, "Peer does not support TLS");
+
+        let handshake = match self.local.as_ref() {
+            Some(local) => {
+                // If ALPN protocols are configured by the endpoint, clone the
+                // configuration and set the protocols. Otherwise, just use the
+                // TLS configuration without modification.
+                //
+                // TODO it would be better to avoid cloning the whole TLS config
+                // per-connection.
+                match alpn {
+                    None => tokio_rustls::TlsConnector::from(local.into()),
+                    Some(AlpnProtocols(protocols)) => {
+                        let mut config: rustls::ClientConfig = local.into().as_ref().clone();
+                        config.alpn_protocols = protocols;
```

Comment:
take it or leave it: might be worth a comment or something making it explicit that this is where the ALPN protocols are actually handed over to `rustls`? I feel like I'm beginning to see the value of making the places where some configuration is handed over to a library really explicit for debugging purposes later on.